### PR TITLE
typing(polys): Add a few type hints in polys

### PR DESCRIPTION
**Pull request by oscarbenjamin**
#### Brief description of what is fixed or changed

A few type hints are add in polys reducing the number of errors reported by pyright from about 10000 to about 2000. The biggest problem apparently was that pyright didn't understand the `@public` decorator.

#### Release Notes

NO ENTRY

### CODE REVIEW

**Comment by oscarbenjamin** [comment]
This reduces the total number of errors from `pyright sympy` from about 40k to about 30k.

File: doc/src/conf.py [hunk 1]
```diff
@@ -530,3 +530,21 @@ def linkcode_resolve(domain, info):
...
+
+
+def resolve_type_aliases(app, env, node, contnode):
+    """Resolve :class: references to our type aliases as :attr: instead."""
+    # A sphinx bug means that TypeVar doesn't work:
+    # https://github.com/sphinx-doc/sphinx/issues/10785
+    if (
+        node["refdomain"] == "py"
+        and node["reftype"] == "class"
+        and node["reftarget"] in ["sympy.utilities.decorator.T"]
```
Comment (sylee957) [inline doc/src/conf.py:542]: I'm okay with the fix. But this looks like we need to add special path to all the typevar references.
I tried building without this extension, but it worked fine though. Maybe you can check the necessity

Comment (oscarbenjamin) [reply doc/src/conf.py:542]: It was failing in CI and also failed locally.

I think I might have found a simpler workaround in gh-27247 though where I used a manual `py:class` reference instead.

**Comment by oscarbenjamin** [comment]
Thanks

**Pull request merged**